### ISSUE TITLE
Wayland support?

### ISSUE THREAD

**Issue by trading-peter** [golang-design/clipboard#6]
Hey quick question. Are there plans to support wayland in the future?
Thank you for you time :)

**Comment by changkun** [comment]
Hi, thanks for writing it up. Yes,  it should be a support to get in, the problem for me is I do not have Wayland to test it properly. VM might be an option but the disk space is quite tight at the moment :)

**Comment by changkun** [comment]
I took a quick look at Wayland systems. It looks like it is fully compatible with X11 and the clipboard package works if we just install libx11-dev.

It seems we don't have to do anything else here.

Did I misunderstand anything?

**Comment by changkun** [comment]
Looks like nothing is needed for Wayland porting. Close.

(clarify if anything was a misunderstanding.)

**Comment by wakatara** [comment]
@changkun 

Can you elaborate a bit more on how this library supports wayland if I am trying to build an app to centralize (and act as a dispatcher - think greenclip on i3/xwindows or the copy/paste and snippets functionality of Alfred on OSX), copies and pastes from sway window manger on wayland. 

Is it just the api for libx11-dev is the same if you are sending messages from wayland to the library, or it's actually *using* xwindows conventions (which for me would be a no go because of the xwindows security model which does not insulate apps.). 

lemme know. Still in the investigative stage of how to build this, so it might just be me not understanding how the mechanics of this works on X11 or Wayland... 

thanks!

**Comment by changkun** [comment]
@wakatara Thanks for the inquiry. In fact, I am still suspect the support regarding Wayland support. As I remember from the last time, I tested the demo cli [`gclip`](https://github.com/golang-design/clipboard/tree/main/cmd/gclip) on Wayland (Switched to Wayland on an X11 Ubuntu). After I installed the `libx11-dev`, the gclip is able to build and could work, although I lack the understanding of if x11 is running on the background or not.

Feel free to let me know if the package could really work from different feedback, as I lack much experience with Wayland systems, but also interested to support it if not yet.

**Comment by microo8** [comment]
Just an idea. Have you considered using [wl-clipboard](https://github.com/bugaevc/wl-clipboard)?

Then just check if we are on wayland:
```
if os.Getenv("WAYLAND_DISPLAY") != "" {
     //call wl-clipboard with os/exec
}
```

**Comment by changkun** [comment]
Yes, I know about wl-clipboard, but it would be an external dependency.

Although it is a temporary solution, I don't have a Wayland environment for testing.

**Comment by microo8** [comment]
So @emersion has pointed me to his blog post: https://emersion.fr/blog/2020/wayland-clipboard-drag-and-drop/
I'll try to do something from it. But first a quick question:

1. Can I introduce a new dependency: [wlroots](https://gitlab.freedesktop.org/wlroots/wlroots) (I try to load it dynamically as the libX11)
2. In the `Init()` function, can I first try to load wlroots and if it fails, then libX11?

**Comment by emersion** [comment]
> Can I introduce a new dependency: [wlroots](https://gitlab.freedesktop.org/wlroots/wlroots) (I try to load it dynamically as the libX11)

This really shouldn't be necessary. wlroots is a library for building *compositors*, clients must not depend on it.

**Comment by emersion** [comment]
> is it the official dev distribution for wayland support?

libwayland is the official C library to connect to a Wayland compositor. It's like libX11 or libxcb.

> What happens if a user is using X but also installed libwayland?

The `WAYLAND_DISPLAY` env variable will be set if the user is inside a Wayland session.

**Comment by microo8** [comment]
So I'll get to it :)

**Comment by laineeded** [comment]
Hello, are there any plans to support wayland soon? or just implementing a temporary solution? thanks!

**Comment by amanhigh** [comment]
Any Update on this ?

**Comment by SR-G** [comment]
This package was working flawlessly on my system with Wayland for quite some time (probably thanks to the X11 bridge), but (and i have no idea why) it stopped working today after the very last Wayland update...

So any news / any update regarding a full Wayland implementation in this very useful library ?

**Comment by SR-G** [comment]
So the only workaround i've been able to find is to ...
- in my main program, stop relying on this library, and expose some REST endpoints (/paste)
- launch a separate command before starting my main program, with : ` wl-paste --type text --watch <path>/clipboard-helper.sh` (the shell is then triggered each time something is copied in the clipboard)
- the shell content is just (in case it would - in a temporary way - be useful for someone) : 

```bash
#!/bin/bash

CONTENT=""
while read -r -N 1 LINE ; do
  CONTENT="${CONTENT}${LINE}"
done < /dev/stdin

if [[ ! -z "$CONTENT" ]] ; then
  curl --silent -L -X POST --data "$CONTENT" http://localhost:<port>/paste
fi
```

This is far away from being a good / proper solution, but it's for now the only working one i can use/think about. 
After several tests, i've not been able in any way to have anymore the "watch" feature working in wayland with this library (since very latest wayland update). The "clipboard" mechanizm in wayland seem to be anyway just a total mess.

**Comment by changkun** [comment]
Folding in #61 (closed as a duplicate of this issue) so the discussion lives in one place.

In a pure Wayland session the X11 backend doesn't just fail to work — it can **crash the host process**. The X selection request is served against an invalid requestor window, producing an async `BadWindow` error on `X_ChangeProperty`; since we don't install a custom Xlib error handler, Xlib's default handler prints and calls `exit(1)`.

Two levels of fix, in increasing scope:

1. **Graceful degradation (small, shippable now):**
   - Install a custom handler via `XSetErrorHandler` (load through `dlsym` like the other symbols) so a `BadWindow`/protocol error is logged or ignored instead of aborting the process. Relevant code: the `XChangeProperty` writes in the `SelectionRequest` handler in [`clipboard_linux.c`](https://github.com/golang-design/clipboard/blob/main/clipboard_linux.c) (~lines 165–182).
   - Optionally detect a Wayland session in `Init()` (`WAYLAND_DISPLAY` set, no usable X display) and return a clear error so callers can handle it instead of crashing.

2. **Native Wayland support (the original ask of this issue):** talk to the Wayland compositor via the data-device protocol (see emersion's notes earlier in this thread).

@ajayd-san offered to look into the crash (item 1) over in #61 — that's a great, self-contained first step and a PR would be very welcome.

**Comment by changkun** [comment]
Update: a design for native Wayland support has been proposed in #109 (`specs/wayland-support.md`).

Approach in brief: a **pure-Go `ext-data-control-v1` client** (with `wlr-data-control` fallback) that speaks the Wayland wire protocol directly — no Cgo, no `libwayland`, no `wl-clipboard` dependency. This uses the data-control protocol specifically because the core `wl_data_device` path requires keyboard focus, which a windowless library doesn't have. `ext-data-control` is now implemented broadly (GNOME ≥ 49 / Mutter, KWin, Sway, Hyprland), so coverage is wide; older compositors fall back to XWayland.

Implementation will begin soon, phased (graceful-fail first, then the data-control read/write/watch paths), each landing with headless-compositor CI. Feedback on the design in #109 is welcome.

**Comment by changkun** [comment]
## Native Wayland support has landed 🎉

Almost four years after this was first opened, `golang.design/x/clipboard` now talks to Wayland natively — **no Cgo, no `libwayland`, no external `wl-clipboard` binary**.

### What shipped
A pure-Go client for the **data-control** protocol, which (unlike the core `wl_data_device`) works **without keyboard focus** — exactly what a windowless library needs:

- Prefers `ext-data-control-v1` (GNOME ≥ 49, KWin) and falls back to `zwlr_data_control_manager_v1` (wlroots: Sway, Hyprland).
- `Init()` auto-selects the Wayland backend when `WAYLAND_DISPLAY` and a data-control manager are present; otherwise it falls back to X11 (via XWayland), so older compositors keep working as before.
- `Read`, `Write` (including the overwrite-notification channel) and event-driven `Watch` all work over the raw Wayland wire protocol, with fd-passing via `SCM_RIGHTS`.
- A regression fix for the related crash (#61, folded in here): a stray X11 protocol error in a pure-Wayland session could `exit(1)` the host process; that no longer happens.

### Tested for real
The full public API is exercised in CI against a **headless Sway** compositor (plus the existing X11 suite under xvfb), so the data-control path is verified end to end on every push — not just built.

### Shipped across these PRs
#109 (design spec), #110, #112, #113, #114, #115, #116.

The design write-up lives at [`specs/wayland-support.md`](https://github.com/golang-design/clipboard/blob/main/specs/wayland-support.md) if you want the protocol details and rationale.

### Thanks 🙏
Huge thanks to everyone who kept this alive and shaped the approach over the years — @microo8 for prototyping and pointing at the right resources, @emersion for the authoritative guidance on Wayland clipboard internals and the [blog post](https://emersion.fr/blog/2020/wayland-clipboard-drag-and-drop/) that the data-control approach is built on, and @wakatara, @bpbpbme, @amanhigh, @SR-G for the reports, use cases, and patience.

Available now on `main`; it'll be in the next tagged release. Please give it a try and open a fresh issue if you hit anything.
